Add audformat.utils.join_schemes()

hagenw opened the pull request:
As we would need to implement this helper function in every data project, I thought it is better to add it directly here.

![image](https://user-images.githubusercontent.com/173624/117971420-2d9a0900-b32a-11eb-8353-c7d8f9aa425e.png)

I also simplified the docstring of `audformat.utils.join_labels()` where we mentioned the code that is now provided by `audformat.utils.join_schemes()`.

frankenjoe commented:
Oh I thought it was already a utility function. So yes, this change makes absolute sense.

hagenw commented:
> Oh I thought it was already a utility function. So yes, this change makes absolute sense.

No, before we focused on just joining the schemes without touching any database.

frankenjoe commented:
Ah sorry, here you actually join the scheme objects. I guess we need to mention that this can have side-effects on the tables.

frankenjoe commented on audformat/core/utils.py:
```diff
@@ -376,6 +360,49 @@ def join_labels(
 
 
+def join_schemes(
+        db1: Database,
```
Since update supports a list of `Database` objects, shouldn't we also allow the same here?

frankenjoe replied:
And would it make sense to actually use the new function in `Database.update()` then?

hagenw replied:
The difference is that `update()` is already assigned to a database.
But in principle you are right, we could replace `db1` and `db2` by a single list `dbs` that can contain one or more databases.
Was that your idea?

frankenjoe replied:
Ah sorry, forget my last comment. We decided earlier that it's safer to let the user do it because of the side-effects it can have.

frankenjoe replied:
> Was that your idea?

yes

frankenjoe replied:
Then it's also in line with `join_labels()`.

hagenw replied:
I changed to a sequence of databases.

frankenjoe commented on audformat/core/utils.py:
```diff
@@ -376,6 +360,43 @@ def join_labels(
 
+def join_schemes(
+        dbs: typing.Sequence[Database],
+        scheme: str,
```
Should we rename to `scheme_id` so that it is more obvious that not a `Scheme` object but an ID is expected?

hagenw replied:
Good idea, I renamed it to `scheme_id`

frankenjoe commented on audformat/core/utils.py:
```diff
@@ -376,6 +360,43 @@ def join_labels(
+        dbs: typing.Sequence[Database],
+        scheme_id: str,
+):
+    r"""Join and update scheme of two databases.
```
Could you please update the docstring as we can now combine the scheme of more than two databases.

hagenw replied:
done

frankenjoe commented on audformat/core/utils.py:
```diff
@@ -376,6 +360,43 @@ def join_labels(
+          labels: [a, b]
+
+    """
+    labels = join_labels([db.schemes[scheme_id].labels for db in dbs])
```
maybe immediately return if `len(dbs) <= 1`?

hagenw replied:
I don't think it will be any speed issue if the code gets executed, or are you worried that the database object is touched with `replace_labels`?

frankenjoe replied:
Ok, as long as `join_labels` works with an empty list we can also leave it as it is.

hagenw replied:
I added a test for it, it does work.

frankenjoe commented on audformat/core/utils.py:
```diff
@@ -376,6 +360,43 @@ def join_labels(
+    with the joined one.
+
+    This might be useful,
+    if you want to combine both databases
```
```suggestion
    if you want to combine databases
```

frankenjoe commented:
> I guess we need to mention that this can have side-effects on the tables.

What do you think, should we maybe add a sentence similar to https://github.com/audeering/audformat/blob/a7ebb6c09815bf522c5a655337d17825443a4e93/audformat/core/scheme.py#L212

hagenw commented:
> > I guess we need to mention that this can have side-effects on the tables.
> 
> What do you think, should we maybe add a sentence similar to
> 
> https://github.com/audeering/audformat/blob/a7ebb6c09815bf522c5a655337d17825443a4e93/audformat/core/scheme.py#L212
> 
> ?

I changed the docstring to:

```rst
    This joins the given scheme of several databases
    using :func:`audformat.utils.join_labels`
    and replaces the scheme in each database
    with the joined one.
    The dtype of all :class:`audformat.Column` objects
    that reference the scheme in the databases
    will be updated.
    Removed labels are set to ``NaN``.

    This might be useful,
    if you want to combine databases
    with :meth:`audformat.Database.update`.
```

frankenjoe commented:
> I changed the docstring to:

Great, thanks! Looks all good now.
